include/graphblas/algorithms/triangle_count.hpp

```diff
@@ -0,0 +1,249 @@
+			{ TriangleCountAlgorithm::Cohen, "Cohen" },
+			{ TriangleCountAlgorithm::Sandia_TT, "Sandia_TT" }
+		};
+
+		template<
+			class Semiring, class MulMonoid, class SumMonoid,
+			Descriptor descr_mxm = descriptors::no_operation,
+			Descriptor descr_ewa = descriptors::no_operation,
+			Descriptor descr_reduce = descriptors::no_operation,
+			typename D1, typename RIT1, typename CIT1, typename NIT1,
+			typename D2, typename RIT2, typename CIT2, typename NIT2,
+			typename D3, typename RIT3, typename CIT3, typename NIT3,
+			typename D4, typename RIT4, typename CIT4, typename NIT4,
+			typename D5, typename RIT5, typename CIT5, typename NIT5,
+			typename D6
+		>
+		RC triangle_count_generic(
+			size_t & count,
+			Matrix< D1, config::default_backend, RIT1, CIT1, NIT1 > & MXM_out,
+			const Matrix< D2, config::default_backend, RIT2, CIT2, NIT2 > & MXM_lhs,
+			const Matrix< D3, config::default_backend, RIT3, CIT3, NIT3 > & MXM_rhs,
+			Matrix< D4, config::default_backend, RIT4, CIT4, NIT4 > & EWA_out,
+			const Matrix< D5, config::default_backend, RIT5, CIT5, NIT5 > & EWA_rhs,
+			const D6 div_factor,
+			const Semiring mxm_semiring = Semiring(),
+			const MulMonoid ewiseapply_monoid = MulMonoid(),
+			const SumMonoid sumreduce_monoid = SumMonoid()
+		) {
+			if( ( &MXM_out == &MXM_lhs ) || ( &MXM_out == &MXM_rhs ) ) {
+				return ILLEGAL;
+			}
+
+			RC rc = SUCCESS;
+
+			// Compute MXM_out = Mlhs * Mrhs
+			rc = rc ? rc : mxm< descr_mxm >( MXM_out, MXM_lhs, MXM_rhs, mxm_semiring, Phase::RESIZE );
+			rc = rc ? rc : mxm< descr_mxm >( MXM_out, MXM_lhs, MXM_rhs, mxm_semiring, Phase::EXECUTE );
+
+			// Compute MXM_out .*= EWA_rhs
+			// FIXME: Replace by a foldl( Matrix[in,out], Matrix[in], Monoid ) - not implemented yet
+			// Will then become:
+			// rc = rc ? rc : eWiseApply< descr_ewa >( MXM_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::RESIZE );
+			// rc = rc ? rc : eWiseApply< descr_ewa >( MXM_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::EXECUTE );
+			// Instead of:
+			rc = rc ? rc : eWiseApply< descr_ewa >( EWA_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::RESIZE );
+			rc = rc ? rc : eWiseApply< descr_ewa >( EWA_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::EXECUTE );
+
+			// Compute a sum reduction over <EWA_out> into <count>
+			count = static_cast< size_t >( 0 );
+			rc = rc ? rc : foldl< descr_reduce >( count, EWA_out, sumreduce_monoid );
+
+			// Apply the div_factor to the reduction result
+			count /= div_factor;
+
+			return rc;
+		}
+
+		/**
+		 * Given a graph, indicates how many triangles are contained within.
+		 *
+		 * @tparam D 				The type of the matrix non-zero values.
+		 *
+		 * @param[out]    count     The number of triangles.
+		 * 						    Any prior contents will be ignored.
+		 * @param[in]     A         The input graph.
+		 * @param[in,out] MXM_out    Buffer matrix with the same dimensions as the input
+		 * 							graph. Any prior contents will be ignored.
```

Comment:
The spec doesn't say `A` and `MXM_out` cannot be the same matrix (that restriction should be added:)